File: tests/test_emulator_pcap_story.py
"""S15 guided ESPShark / PCAP analysis driven through the native viewer."""
import unittest
from emulator_browser import objects, browser_session
import test_emulator_phase4_pcap_deep as deep

class PcapStoryBrowser(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.session=browser_session();cls.browser,cls.url=cls.session.__enter__()
    @classmethod
    def tearDownClass(cls):
        cls.session.__exit__(None,None,None)
    text=deep.PcapDeepBrowser.text
    tick=deep.PcapDeepBrowser.tick
    wait_text=deep.PcapDeepBrowser.wait_text
    click=deep.PcapDeepBrowser.click
    click_label=deep.PcapDeepBrowser.click_label
    open_capture=deep.PcapDeepBrowser.open_capture
    close_analysis=deep.PcapDeepBrowser.close_analysis

    def setUp(self):
        self.context=self.browser.new_context()
        self.addCleanup(self.context.close)
        self.page=self.context.new_page();self.errors=[]
        self.page.on('pageerror',lambda e:self.errors.append(str(e)))

    def begin(self,story_id,rotation=0):
        self.page.goto(f'{self.url}/?rotation={rotation}')
        self.page.wait_for_function('globalThis.emulator?.measurements.frames>2')
        self.page.evaluate("globalThis.unavailable=[];addEventListener('emulator-unavailable',e=>unavailable.push(e.detail))")
        self.page.locator('#queue2-story').select_option(story_id)
        self.page.locator('#guide-queue2').click()
        self.page.evaluate("async id=>{const {pcapStories}=await import('./pcap-story.mjs');globalThis.pcStory=pcapStories.find(e=>e.id===id).create().story;}",story_id)
        self.page.locator('#pcap-examples').click();self.wait_text('example-http-dns-icmp.pcap')

    def stage(self,n):
        self.page.wait_for_function('(n)=>document.querySelector("#guide-title").textContent===(n===pcStory.steps.length?pcStory.completionTitle:pcStory.steps[n].title)',arg=n,timeout=9000)

    def done(self):
        self.page.wait_for_function('document.querySelector("#guide-title").textContent===pcStory.completionTitle',timeout=9000)
        self.assertIn('complete',self.page.locator('#guide-progress').inner_text())
        self.assertEqual(self.errors,[]);self.assertEqual(self.page.evaluate('unavailable'),[])

    # The deep submenu walks drive the native scrolling viewer at the default
    # rotation, matching Phase 4.3 coverage. Guide behaviour is rotation-invariant
    # (it reads native text): the four-rotation evidence is the invalid/recovery
    # walk below, and every guide state machine is covered per rotation in the
    # logic suite test_emulator_pcap_story.mjs.
    def test_analysis_walks_every_submenu(self):
        self.begin('espshark-analysis')
        self.open_capture();self.stage(1)
        self.click('pcap_viewer_render_capture_page.map_btn');self.wait_text('ESPShark Network Topology');self.stage(2)
        self.click('pcap_viewer_map_cb.close')
        self.click('pcap_viewer_render_capture_page.devices_btn');self.wait_text('Local Devices (tap to filter)');self.stage(3)
        remote=next(o['text'] for o in objects(self.page) if o['text'].startswith('REMOTE ('))
        self.click_label(remote);self.wait_text('Remote Endpoints (tap to filter)');self.stage(4)
        self.close_analysis()
        self.click('pcap_viewer_render_capture_page.connections_btn');self.wait_text('Connections (tap FILTER or FOLLOW)');self.stage(5)
        self.close_analysis()
        self.click('pcap_viewer_render_capture_page.protocols_btn');self.wait_text('Top Application Protocols');self.stage(6)
        self.close_analysis();self.done()

    def test_investigate_packet_health_fqdn_export(self):
        self.begin('espshark-investigate')
        self.open_capture();self.stage(1)
        self.click(':packet-1/packet');self.wait_text('Packet #1 details');self.stage(2)
        self.click('pcap_viewer_packet_detail_cb.close_btn')
        self.click('pcap_viewer_render_capture_page.health_btn');self.wait_text('ESPShark Offline Investigation')
        for mode in ('TIMELINE','BASELINE','INTEL'):self.click_label(mode)
        self.wait_text('Offline intelligence:');self.stage(3)
        self.close_analysis()
        self.click('pcap_viewer_render_capture_page.state--fqdn_btn');self.wait_text('FQDN ON');self.stage(4)
        self.click('pcap_viewer_render_capture_page.tools_btn');self.wait_text('ESPShark Cache & Export');self.stage(5)
        self.click_label('CLOSE');self.done()

    def test_extract_objects_preview_and_hash(self):
        self.begin('espshark-extract')
        self.open_capture();self.stage(1)
        self.click('pcap_viewer_render_capture_page.objects_btn');self.wait_text('Extracted Objects (1)')
        self.assertIn('HTTP 200',self.text());self.stage(2)
        self.click_label('PREVIEW');self.wait_text('Offline synthetic example.')
        self.assertIn('SHA-256',self.text());self.stage(3)
        self.click('pcap_viewer_show_summary_popup.close_btn')
        if 'Extracted Objects' in self.text():self.close_analysis()
        self.done()

    def test_invalid_capture_then_recovery_all_rotations(self):
        # Rotation-independence evidence for the guide across all four rotations.
        for rotation in range(4):
            with self.subTest(rotation=rotation):
                self.begin('espshark-invalid',rotation)
                self.click('example-invalid-truncated.pcap/open');self.wait_text('Truncated PCAP');self.stage(1)
                self.open_capture();self.done()

if __name__=='__main__':unittest.main(verbosity=2)

File: tests/test_emulator_pcap_story.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import {pcapStories} from '../tools/ui_emulator/web/pcap-story.mjs';

const entry=id=>pcapStories.find(x=>x.id===id);
const base={tab:0,captureOpen:false,map:false,devices:false,endpoints:false,connections:false,
 protocols:false,objects:false,tools:false,packet:false,investigation:false,http:false,http200:false,
 preview:false,hash:false,intel:false,fqdnOn:false,invalid:false,export:false,analysisOpen:false,capturePage:false};
// Mirror the reader: capturePage is captureOpen with no analysis view open.
const view=(open,...flags)=>{const s={...base,captureOpen:open};for(const f of flags)s[f]=true;
 s.analysisOpen=['map','devices','endpoints','connections','protocols','objects','tools','packet','investigation','preview','hash'].some(k=>s[k]);
 s.capturePage=s.captureOpen&&!s.analysisOpen;return s;};

test('S15 exposes analysis, investigate, extract and invalid variants',()=>{
 assert.deepEqual(pcapStories.map(x=>x.id),['espshark-analysis','espshark-investigate','espshark-extract','espshark-invalid']);
 for(const e of pcapStories){const g=e.create();g.start({});assert.ok(g.story.steps.length>=2);
  g.accept({...view(true),epoch:g.epoch-1});assert.equal(g.step,0,'stale epoch rejected');g.leave();assert.equal(g.active,false);}
});

test('analysis walks every inspection submenu then returns',()=>{
 const g=entry('espshark-analysis').create();g.start({});const emit=s=>g.accept({...s,epoch:g.epoch});
 emit(view(false));assert.equal(g.step,0,'needs an open capture');
 emit(view(true));assert.equal(g.step,1);
 emit(view(true,'map'));assert.equal(g.step,2);
 emit(view(true,'devices'));assert.equal(g.step,3);
 emit(view(true,'endpoints'));assert.equal(g.step,4);
 emit(view(true,'connections'));assert.equal(g.step,5);
 emit(view(true,'protocols'));assert.equal(g.step,5,'protocols step needs HTTP');
 emit(view(true,'protocols','http'));assert.equal(g.step,6);
 emit(view(true));assert.equal(g.complete,true);
});

test('investigate covers packet, health, FQDN and export',()=>{
 const g=entry('espshark-investigate').create();g.start({});const emit=s=>g.accept({...s,epoch:g.epoch});
 emit(view(true));assert.equal(g.step,1);
 emit(view(true,'packet'));assert.equal(g.step,2);
 emit(view(true,'investigation'));assert.equal(g.step,2,'needs intel');
 emit(view(true,'investigation','intel'));assert.equal(g.step,3);
 emit(view(true,'fqdnOn'));assert.equal(g.step,4);
 emit(view(true,'tools'));assert.equal(g.step,4,'needs export label');
 emit(view(true,'tools','export'));assert.equal(g.step,5);
 emit(view(true));assert.equal(g.complete,true);
});

test('extract reaches objects, preview and the SHA-256 hash',()=>{
 const g=entry('espshark-extract').create();g.start({});const emit=s=>g.accept({...s,epoch:g.epoch});
 emit(view(true));assert.equal(g.step,1);
 emit(view(true,'objects'));assert.equal(g.step,1,'needs HTTP 200');
 emit(view(true,'objects','http200'));assert.equal(g.step,2);
 emit(view(true,'objects','preview'));assert.equal(g.step,2,'needs hash');
 emit(view(true,'objects','preview','hash'));assert.equal(g.step,3);
 emit(view(true));assert.equal(g.complete,true);
});

test('invalid capture guides recovery to a valid one',()=>{
 const g=entry('espshark-invalid').create();g.start({});const emit=s=>g.accept({...s,epoch:g.epoch});
 emit(view(false));assert.equal(g.step,0);
 emit({...view(false),invalid:true});assert.equal(g.step,1);
 emit(view(true));assert.equal(g.complete,true);
});

test('losing the capture pauses an analysis story',()=>{
 const g=entry('espshark-analysis').create();g.start({});const emit=s=>g.accept({...s,epoch:g.epoch});
 emit(view(true));emit(view(true,'map'));assert.equal(g.step,2);
 emit(view(false));assert.equal(g.step,2);assert.match(g.hint,/Reopen/);
});
